### PR TITLE
Update cache to ignore values over 250 characters (D1118)

### PR DESCRIPTION
**Pull request by droberts-ctrlo**
This code checks the `%values` hash and does not insert the value into the cache should there be any values over 250 characters inserted into the field.

### CODE REVIEW

**Comment by pawelwlod** [comment]
> This code checks the `%values` hash and does not update insert the value into the cache should there be any values over 250 characters inserted into the field.

What do you mean by "update insert the value"?

**Comment by droberts-ctrlo** [comment]
> > This code checks the `%values` hash and does not update insert the value into the cache should there be any values over 250 characters inserted into the field.
> 
> What do you mean by "update insert the value"?

Fixed

File: t/007_code_cache.t [hunk 1]
```diff
@@ -11,6 +11,8 @@ use GADS::Schema;
 use lib 't/lib';
 use Test::GADS::DataSheet;
 
+use Data::Dump qw(pp);
```
Comment (pawelwlod) [inline t/007_code_cache.t]: Is this a debugging dependency?

Comment (droberts-ctrlo) [reply t/007_code_cache.t]: It is - I shall remove!

File: lib/GADS/Datum/Code.pm [hunk 2]
```diff
@@ -162,6 +162,8 @@ sub write_cache
     my $formatter = $self->schema->storage->datetime_parser;
 
     my @values = sort @{$self->value} if defined $self->value->[0];
+    # If the length is greater than 250 characters do not insert it to the cache table
+    return \@values if grep { $_ > 250 } map { length($_) } @values;
```
Comment (abeverley) [inline lib/GADS/Datum/Code.pm]: Unfortunately returning here means that the calculated value is not written to the database field at all. This means that any searches and rendering with prefetching will not work.

File: t/007_code_cache.t [hunk 3]
```diff
@@ -181,4 +181,24 @@ is_deeply([map $_->value_int, $urs->all], $expected, "Correct values for cached
...
+    my $expected_calc_val = "This is a really long string, longer than before, to test caching of long strings should fail gracefully and not put any extra values in the cache, this is to test the edge case handling of long strings in cached unique values because they break the cache in some really weird ways sometimes";
+    $record->clear;
+    $record->find_current_id(1);
+    $calc1->code("
+        function evaluate (L1string1)
+            return L1string1
+        end
+    ");
+    $calc1->return_type('string');
+    $calc1->write;
```
Comment (abeverley) [inline t/007_code_cache.t:197]: Might need to clear the layout here for the new field (see other examples) although no problem if it's working.

File: t/007_code_cache.t [hunk 4]
```diff
@@ -181,4 +181,24 @@ is_deeply([map $_->value_int, $urs->all], $expected, "Correct values for cached
...
+    $record->fields->{$string1->id}->set_value($expected_calc_val);
+    $record->write(no_alerts => 1);
+
+    is_deeply([map $_->value_text, $urs->all], [qw/Bar7 Bar8/], "Correct values for cached table after long string write");
```
Comment (abeverley) [inline t/007_code_cache.t]: Whilst this specifically is correct, we should also test whether the (now unused) calc unique fields are deleted.

We should also test whether the value in the `Calcval` table (probably better tested in `t/007_code.t`)

File: t/007_code_cache.t [hunk 5]
```diff
@@ -181,4 +181,24 @@ is_deeply([map $_->value_int, $urs->all], $expected, "Correct values for cached
...
+
+    my $calc_val = $record->fields->{$string1->id}->as_string;
```
Comment (abeverley) [inline t/007_code_cache.t]: This should be the calc field not the string field.

File: t/007_code_cache.t [hunk 6]
```diff
@@ -181,4 +181,24 @@ is_deeply([map $_->value_int, $urs->all], $expected, "Correct values for cached
...
+    is($calc_val, $expected_calc_val, "Correct value returned from calc after long string write");
```
Comment (abeverley) [inline t/007_code_cache.t]: I think we also need to add a test for `GADS::Records` searching on the long calc value (probably better placed in `t/007_code.t`. I think that will fail with the code as is (for the reasons above).

**Review by abeverley — COMMENTED** [review]
Thanks Dave, good work on getting the test added. A bit more to do though I'm afraid...

**Comment by abeverley** [comment]
Thanks Dave, this is great. Just 2 very minor things before merging: add a brief comment to the new test explaining what it is testing (I know I haven't always done this, and I've seen the problems it creates!); plus, remove the Data::Dumper from the included libraries.
